pkg/reconciler/pipelinerun/resources/pipelinerunstate.go

```diff
@@ -307,7 +307,7 @@ func (facts *PipelineRunFacts) GetFinalTasks() PipelineRunState {
 	finalCandidates := sets.NewString()
 	// check either pipeline has finished executing all DAG pipelineTasks
 	// or any one of the DAG pipelineTask has failed
-	if facts.checkDAGTasksDone() {
+	if facts.checkDAGTasksDone() || (!facts.IsGracefullyStopped() || !facts.IsGracefullyCancelled()) {
```

Comment:
According to the `checkDAGTasksDone` docstring, it checks if all the DAG tasks are done executing, including succeeded, failed, and skipped. Shouldn't a task that has failed both retries be considered done? It seems like there might be a bug in that function rather than this one.